server/apps/pa_img/pa_img_datalayer.py

```diff
@@ -0,0 +1,265 @@
+        if self._token:
+            fields['token'] = self._token
+
+        url = self._app.config['PAIMG_SEARCH_URL'] + '/meta/{}'.format(_id)
+        r = self._http.request_encode_url('GET', url, fields=fields, headers=self._headers)
+        doc = json.loads(r.data.decode('UTF-8'))
+        self._parse_doc(doc)
+
+        if 'fetch_endpoint' in doc:
+            del doc['fetch_endpoint']
+
+        if doc[ITEM_TYPE] == CONTENT_TYPE.PICTURE:
+            if 'original' in doc.get('renditions', None):
+                url = doc['renditions']['original']['href']
+                mime_type = 'image/jpeg'
+            else:
+                raise FileNotFoundError
+        else:
+            raise NotImplementedError
+
+        r = self._http.request_encode_url('GET', url, fields=fields, headers=self._headers)
+        out = BytesIO(r.data)
+        file_name, content_type, metadata = process_file_from_stream(out, mime_type)
+
+        inserted = []
+
+        try:
+            logger.debug('Going to save media file with %s ' % file_name)
+            out.seek(0)
+            file_id = self._app.media.put(out, filename=file_name, content_type=content_type, metadata=None)
+            doc['mimetype'] = content_type
+            doc['filemeta'] = decode_metadata(metadata)
+            # set the version created to now to bring it to the top of the desk, images can be quite old
+            doc['versioncreated'] = utcnow()
+            inserted = [file_id]
+            file_type = content_type.split('/')[0]
+            rendition_spec = self._app.config['RENDITIONS']['picture']
+
+            renditions = generate_renditions(out, file_id, inserted, file_type,
+                                             content_type, rendition_spec,
+                                             self.url_for_media, insert_metadata=False)
+            doc['renditions'] = renditions
+        except Exception as io:
+            logger.exception(io)
+            for file_id in inserted:
+                delete_file_on_error(doc, file_id)
+
+            raise SuperdeskApiError.internalError('Generating renditions failed')
+
+        return doc
+
+    def url_for_media(self, media_id):
+        return url_for('upload_raw.get_upload_as_data_uri', media_id=media_id,
+                       _external=True, _schema=self._app.config['URL_PROTOCOL'])
+
+    def find_list_of_ids(self, resource, ids, client_projection=None):
+        raise NotImplementedError
+
+    def insert(self, resource, docs, **kwargs):
+        raise NotImplementedError
+
+    def update(self, resource, id_, updates, original):
+        raise NotImplementedError
+
+    def update_all(self, resource, query, updates):
+        raise NotImplementedError
+
+    def replace(self, resource, id_, document, original):
+        raise NotImplementedError
+
+    def remove(self, resource, lookup=None):
+        raise NotImplementedError
+
+    def is_empty(self, resource):
+        raise NotImplementedError
```

Comment:
Looking at all those `raise NotImplementedError` method bodies, is the `PaImgDatalayer` meant to be an abstract class? If so, it would probably make sense to set its metaclass to `ABCmeta` and mark its abstract method with the `@abstractmethod` decorator to prevent instantiation?
(instead of raising the `NotImplementedError` only after an attempt has been made to use a non-overriden abstract method).